cmd/common/oneshot.go

```diff
@@ -0,0 +1,22 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package common
+
+import "go.uber.org/fx"
+
+func shutdown(sd fx.Shutdowner) {
+	sd.Shutdown()
+}
+
+// OneShot is an fx.Option that calls the given function, using fx.Invoke to
+// load its arguments, and causes the app to shut down when the function
+// returns.  It's used for one-shot agent subcommands like `agent status`.
+func OneShot(f interface{}) fx.Option {
+	return fx.Options(
+		fx.Invoke(f),
```

Comment:
`Invoke` is called before any hooks `fx.Hook`: see https://go.dev/play/p/Bble3f8A0Dc.
It means `f` cannot assume any of its dependencies are ready (as a dependency can contain hooks). 

As a workaround, the code of `f` can be wrapped in a `OnStart` hook but it let me think `fx` was designed for micro services.

Reply:
Yes, that's correct.  In general, I think that doesn't hurt anything, but you're right that it's a mistake here: OneShot should be scheduling `f` to run in a Hook.  In general, use of fx.Invoke is probably going to be unusual.